modules/gateway/peers.go

```diff
@@ -437,10 +523,17 @@ func (g *Gateway) managedConnect(addr modules.NetAddress) error {
 		conn.Close()
 		return err
 	}
-	if build.VersionCmp(remoteVersion, handshakeUpgradeVersion) < 0 {
-		err = g.managedConnectOldPeer(conn, remoteVersion, addr)
-	} else {
+
+	if build.VersionCmp(remoteVersion, sessionHandshakeUpgradeVersion) >= 0 {
+		wantConnect := true
+		err = connectSessionHandshake(conn, g.id, wantConnect)
+		if err == nil {
+			err = g.managedConnectNewPeer(conn, remoteVersion, addr)
+		}
```

Comment:
hmm. This is ugly. We really try to avoid writing `if err == nil` because `err != nil` is so much more common. I would rather this look like:
```go
if build.VersionCmp(remoteVersion, sessionHandshakeUpgradeVersion) >= 0 {
    err = g.managedConnectv120Peer(conn, remoteVersion, addr)
} else if build.VersionCmp(remoteVersion, handshakeUpgradeVersion) >= 0 {
    err = g.managedConnectv100Peer(conn, remoteVersion, addr)
} else {
    err = g.managedConnectOldPeer(conn, remoteVersion, addr)
}
```